### DISCUSSION TITLE
IKEA AQI sensor - Is this JSON format supported

### DISCUSSION THREAD

**Discussion by gm1ymi** [bellrichm/WeeWX-MQTTSubscribe#170] [Q&A] [answered]
Here is the MQTT format. I am only needing the first section - "pm25":2 
Client mosqsub|21219-k1m received PUBLISH (d0, q0, r0, m0, 'esp8266-vindriktning-particle-sensor/VINDRIKTNING-CFCDF0/state', ... (79 bytes))
esp8266-vindriktning-particle-sensor/VINDRIKTNING-CFCDF0/state {"pm25":2,"wifi":{"ssid":"guest","ip":"192.168.1.152","rssi":-72}}

MQTTSubscribe is complaining about the wifi data (which I don't want anyway).

in weewx.conf I have the following:

 [[[esp8266-vindriktning-particle-sensor/VINDRIKTNING-CFCDF0/state]]]
        ignore = True  

           [[[pm25]]]
           name = pm25
           ignore = False

If I remove the 'ignore = True' syslog is filled with errors like this. relating to the wifi portion:

Nov  1 15:11:45 k1m weewx[20269] ERROR user.MQTTSubscribe: (Service) **** MessageCallbackProvider Traceback (most recent call last):#012  File "/usr/share/weewx/user/MQTTSubscribe.py", line 1300, in _convert_value#012    return conversion_func['compiled'](value)#012  File "<string>", line 1, in <lambda>#012  File "/usr/share/weewx/weeutil/weeutil.py", line 1533, in to_float#012    return float(x) if x is not None else None#012ValueError: could not convert string to float: 'guest'#012#012During handling of the above exception, another exception occurred:#012#012Traceback (most recent call last):#012  File "/usr/share/weewx/user/MQTTSubscribe.py", line 1492, in _on_message_json#012    (fieldname, value) = self._update_data(fields, fields_conversion_func, lookup_key, data_flattened[key], unit_system)#012  File "/usr/share/weewx/user/MQTTSubscribe.py", line 1259, in _update_data#012    value = self._convert_value(fields, default_field_conversion_func, orig_name, orig_value)#012  File "/usr/share/weewx/user/MQTTSubscribe.py", line 1306, in _convert_value#012    % (field, value, conversion_func['source'], exception))#012user.MQTTSubscribe.ConversionError: Failed converting field wifi_ssid with value Surveillance Van#2 using 'lambda x: to_float(x)' with reason could not convert string to float: 'guest'.

If I leave the 'ignore = True' in then I get no errors, but weewx does not seem to parse the "pm25" data. As it does not show in syslog or on weewx webpage.

I am going to see if I can drop the extra data from the AQI sensor, but it would be good to be able to handle it in MQTTSubscribe too.
Any pointers to handle this "mixed" json format.
Thanks

**Comment by bellrichm** [comment]
I think you are missing one set of `[]` around the pm25. Try adding one for a total of 4
```
[[[esp8266-vindriktning-particle-sensor/VINDRIKTNING-CFCDF0/state]]]
    ignore = True
    [[[[pm25]]]]
       name = pm25
       ignore = False
```

**Reply by gm1ymi** [reply]
Oh, I see that now, thanks.
I added the 4th [ ] and still nothing showing in weewx.
In syslog I have the following regarding to MQTT. So it looks like it's subscribing... I thnk... New to MQTT ... so learning as I stumble along.

Nov  1 20:59:20 k1ymi weewx[29055] INFO user.MQTTSubscribe: (Service) Disconnected with result code 0
Nov  1 20:59:21 k1ymi weewx[31290] INFO user.MQTTSubscribe: (Service) Version is 2.2.2
Nov  1 20:59:21 k1ymi weewx[31290] INFO user.MQTTSubscribe: (Service) Log level: 0
Nov  1 20:59:21 k1ymi weewx[31290] INFO user.MQTTSubscribe: (Service) Log debug setting: 0
Nov  1 20:59:21 k1ymi weewx[31290] INFO user.MQTTSubscribe: (Service) Log console: False
Nov  1 20:59:21 k1ymi weewx[31290] INFO user.MQTTSubscribe: (Service) Log file: None
Nov  1 20:59:21 k1ymi weewx[31290] INFO user.MQTTSubscribe: (Service) message_callback_provider_name is user.MQTTSubscribe.MessageCallbackProvider
Nov  1 20:59:21 k1y weewx[31290] INFO user.MQTTSubscribe: (Service) clientid is MQTTSubscribe-1005
Nov  1 20:59:21 k1y weewx[31290] INFO user.MQTTSubscribe: (Service) client_session is True
Nov  1 20:59:21 k1 weewx[31290] INFO user.MQTTSubscribe: (Service) host is localhost
Nov  1 20:59:21 k1y weewx[31290] INFO user.MQTTSubscribe: (Service) port is 1883
Nov  1 20:59:21 k1y weewx[31290] INFO user.MQTTSubscribe: (Service) keepalive is 60
Nov  1 20:59:21 k1yweewx[31290] INFO user.MQTTSubscribe: (Service) username is None
Nov  1 20:59:21 k1y weewx[31290] INFO user.MQTTSubscribe: (Service) min_delay is 1
Nov  1 20:59:21 k1y weewx[31290] INFO user.MQTTSubscribe: (Service) max_delay is 120
Nov  1 20:59:21 k1yweewx[31290] INFO user.MQTTSubscribe: (Service) password is set
Nov  1 20:59:21 k1y weewx[31290] INFO user.MQTTSubscribe: (Service) Archive topic is None
Nov  1 20:59:21 k1yweewx[31290] INFO user.MQTTSubscribe: (Service) binding is loop
Nov  1 20:59:21 k1yweewx[31290] INFO user.MQTTSubscribe: (Service) Waiting for MQTT connection.
Nov  1 20:59:21 k1yweewx[31290] INFO user.MQTTSubscribe: (Service) Connected with result code 0
Nov  1 20:59:21 k1yweewx[31290] INFO user.MQTTSubscribe: (Service) Connected flags {'session present': 0}
Nov  1 20:59:21 k1y weewx[31290] INFO user.MQTTSubscribe: (Service) Subscribing to esp8266-vindriktning-particle-sensor/VINDRIKTNING-CFCDF0/state has a mid 1 and rc 0
Nov  1 20:59:21 k1yweewx[31290] INFO user.MQTTSubscribe: (Service) Subscribed to mid: 1 is size 1 has a QOS of 0
Nov  1 20:59:22 k1y weewx[31290] INFO user.MQTTSubscribe: (Service) MQTT initialization complete.

**edit** The next lines are from a later grep of syslog than the inital post above.
Nov  1 21:14:57 k1y weewx[31290] INFO user.MQTTSubscribe: (Service) TopicManager ignoring record outside of interval 1667351697.000000 1667351697.000000 1667351696.992496 'dateTime': '1667351696.9924955', 'pm25': '4.0', 'usUnits': '1'
Nov  1 21:38:58 k1y weewx[31290] INFO user.MQTTSubscribe: (Service) TopicManager ignoring record outside of interval 1667353138.000000 1667353138.000000 1667353137.959308 'dateTime': '1667353137.9593084', 'pm25': '3.0', 'usUnits': '1'
Nov  1 21:41:58 k1y weewx[31290] INFO user.MQTTSubscribe: (Service) TopicManager ignoring record outside of interval 1667353318.000000 1667353318.000000 1667353317.947131 'dateTime': '1667353317.9471307', 'pm25': '3.0', 'usUnits': '1'
Nov  1 21:42:28 k1y weewx[31290] INFO user.MQTTSubscribe: (Service) TopicManager ignoring record outside of interval 1667353348.000000 1667353348.000000 1667353347.947076 'dateTime': '1667353347.9470763', 'pm25': '3.0', 'usUnits': '1'
Nov  1 22:13:59 k1y weewx[31290] INFO user.MQTTSubscribe: (Service) TopicManager ignoring record outside of interval 1667355239.000000 1667355239.000000 1667355238.913908 'dateTime': '1667355238.9139078', 'pm25': '3.0', 'usUnits': '1'

**Comment by bellrichm** [comment]
Looks like something strange/bug in MQTTSubscribe’s datetime checking. This should workaround it.
```
    [[topics]
        # When True, the MQTT datetime will be not be checked that is greater than the last packet processed.
        # Default is False.
        # Only used by the service.
        ignore_start_time = True

        # When the True, the MQTT data will continue to be processed even if its datetime is greater than the packet's datetime.
        # Default is False.
        # Only used by the service.
        ignore_end_time = True
```

**Reply by gm1ymi** [reply]
Yes, that seems to have resolved the datetime error. 
Weewx has been running about an hour and no errors noted for MQTTSubscribe in syslog.
syslog shows 'MQTT initialization complete.' then nothing since the restart. No MQTT data from the device.
Is there anything else in weewx I need to do to get this to start graphing? I *think* once it ingested the data it would start producing the graph.... but I maybe totally wrong.
**Edit** --> I am going to try the Tasmota ESP8266 firmware on the sensor and see if that makes a difference. I will follow up.
Thanks for all the help!

**Comment by bellrichm** [comment, marked as answer]
I know you have switched things up, but I think in the original setup you want is to set `name = pm2_5`
reference: https://github.com/weewx/weewx/blob/master/bin/schemas/wview_extended.py
That should get it into the db. Then it is dependent on the skin you are using to graph it. Not sure, but I think the seasons skin wil graph it…

**Reply by gm1ymi** [reply]
Great timing!
I had just switched back to the original ESP8266 code (https://github.com/Hypfer/esp8266-vindriktning-particle-sensor)
With the correction to the name field it is now graphing, and there are no errors in syslog. Thanks.
For clarity of the solution here is my configuration.

[[topics]]
        # Units for MQTT payloads without unit value.
        # Valid values: US, METRIC, METRICWX
        # Default is: US
        unit_system = US
        # When True, the MQTT datetime will be not be checked that is greater than the last packet processed.
        # Default is False.
        # Only used by the service.
        ignore_start_time = True

        # When the True, the MQTT data will continue to be processed even if its datetime is greater than the packet's datetime.
        # Default is False.
        # Only used by the service.
        ignore_end_time = True

        [[[esp8266-vindriktning-particle-sensor/VINDRIKTNING-CFCDF0/state]]]
           ignore = True

           [[[[pm25]]]]
           name = pm2_5
           ignore = False

**Reply by bellrichm** [reply]
Good to hear it is working. And thanks posting your configuration, hopefully it will help someone out.
